Fix comparison of changed files

pascalkuthe opened the pull request:
This PR is a major refactor of the core diff logic to fix #26.
The crate completely removes any textual diffing that was previously performed using `imara-diff`.
I verified that `imara-diff` function correctly.
The problem is that `imara-diff` (and diffs in general) are intended to provide linear editing sequences.
That is a set of edits to transform file A into file B **without any jumps**.
However, for `crates-index-diff` we actually don't care about this property, because we only care about how any given
checksum changed (removed/added/yanked).
This means that `imara-diff` provides unideal diffs for this usecase.
For example if a version X is moved from the end of the file to the start of the file imara diff provides to changes:
add X at line 0
remove X at last line

That means that any `remove` changes provided by `imara-diff` had to be ignored for correct functionality.
Ignoring removed lines works when crates are never deleted.
However, if comparing two (far apart) commits between which a crate was deleted and later reclaimed by someone else.
Then we need to mark any removed checksum as deleted.

This PR changes the diff logic to perform a simple unordered `HashMap<Checksum, CrateVersion>` comparison.
The details are a bit tricky to avoid parsing every line (and to avoid storing the checksum twice).
However, the code is functionally equivalent to the following s (non compiling pseudocode, some details left out):

``` rust
let old_versions: HashMap<Checksum, CrateVersion> = diff.old.lines().map(|line| parse_crate_version(line));
let new_versions: HashMap<Checksum, CrateVersion> = diff.new.lines().map(|line| parse_crate_version(line));
for new_version in new_versions.difference(old_versions){
  self.changes.push(Change::Added(new_version))
}
for old_version in old_versions.difference(new_versions){
  self.changes.push(Change::Deleted(old_version))
}
for (old_version, new_version) in old_versions.intersection(new_versions){
  let change = match (old_version.yanked, new_version.yanked) {
      (true, false) => Change::Unyanked(new_version),
      (false, true) => Change::Yanked(new_version),
      _ => continue,
  };
  self.changes.push(change)
}
```

In the process I discovered some ambiguities in the API of `Change` around yanked crates.
To address that I have added two new variants to that enum:

``` rust
Unyanked(CrateVersion)
AddedAndYanked(CrateVersion)
```

This ensures that downstream crates can always tell whether a crate
was yanked/unyanked/added without any extra operations.

Another API change:

Due to the refactor of the diff algorithm, changes are now naturally grouped
by crate. That means additional sorting inside `crates-index-diff` is not required to achieve that property.
To improve performance the sorting by crate-name/version has been removed.
If sorting is required by downstream crates, it is trivial for them to call `changes.sort_by` themselves.
This change is a nice performance boost for any crates that don't care about the order of changes (or require a different sorting then the one that was used by `crates-index-diff`).
Note that the order of the changes is completely non-deterministic (except for being grouped by crate).
So if any downstream crate (or a test inside this crate) requires a deterministic order, it must sort the vector.


This PR fixes the bug discussed in #26 where `crates-index-diff` would report
a different set of changes when splitting the diff across multiple commits.
This PR also fixed the `baseline` test which previously forget to include the `HEAD` commit (and only included ancestors).

pascalkuthe committed:
perform an unordered comparison instead of using a linear edit-sequence

Byron committed:
refactor (#26)

- typos and form
- improve docs and docs consistency.

Byron commented on src/index/diff/mod.rs:
```diff
@@ -164,6 +164,16 @@ impl Index {
     /// Similar to `changes()`, but requires `from` and `to` objects to be provided. They may point
     /// to either `Commit`s or `Tree`s.
+    ///
+    /// # Returns
```
I really love these docs, thank you!

Byron commented on src/types.rs:
```diff
@@ -59,6 +71,20 @@ impl Change {
+            Change::Added(v)
+            | Change::Unyanked(v)
+            | Change::AddedAndYanked(v)
+            | Change::Yanked(v) => slice::from_ref(v),
```
`slice::from_ref()` is a little gem that I learned about recently enough to still be excited about it :).

pascalkuthe replied:
I love this function! It's perfect for enapsulating `one or many` without another enum.
I found this little gem in the cranelift code, at some point

pascalkuthe committed:
add baseline tests that steps trough each commit individually

Byron committed:
improve baseline tests to be more practical

- don't run them unless it's a release build
- make the existence of atomic tests more obvious with `make`
- share code between both baseline variants
- some output to learn about progress

Byron commented on tests/baseline.rs:
```diff
@@ -16,35 +18,58 @@ fn all_aggregrated_diffs_equal_latest_version(
+                                versions.insert(v.checksum.to_owned());
+                            }
+                            Unyanked(v) | Yanked(v) => {
+                                // yanked/unyanked crates must be part of the index
```
Loving this added assertion for extra sanity checking :).

Byron committed:
feat: a `max-performance` feature to tune `git-repository`.

The performance difference is rather drastic at about 2.5x, and
definitely worth having if there is no compatibility issue
due to shared C dependencies in the same binary.

Additionally we setup the makefile to use big object caches
to avoid having to decompress the same object too often, accelerating
the diffing process about 4x, for a total of 10x performance boost.

Byron reviewed (changes requested):
Fantastic work, thanks so much!

Thanks to `git-lfs` I can't push my changes to the PR directly, but have attached the patches here.

[patches.zip](https://github.com/Byron/crates-index-diff-rs/files/10050659/patches.zip) (_Updated_)


Before merging, I'd love if you could try to remove the `unsafe` from the code as it's unclear if these carry any weight without profiling a before and after. From the profile attached here (diffing commit by commit) the time for the actual diffing is marginal at best, and even though bigger steps (i.e. more changes) will result in more work for the differ, I doubt it can stand up to the time it takes to decompress objects. `docs.rs` pulls regularly and there is only ~650 changes per day.

<img width="1163" alt="Screenshot-2022-11-20-at-20 27 12" src="https://user-images.githubusercontent.com/63622/202924065-f6c2473d-b0f0-46d1-aeb6-e04e372f2409.png">

And with that it should be ready to merge.

pascalkuthe committed:
remove unnecessary unsafe code

Byron reviewed (approved):
And it's a wrap! Thanks so much for all the effort that you put into this project, it's on an entirely new level thanks to it and I think it might finally be 'done' (despite #30 which might be implemented one fine day or maybe never).
